gwpy/timeseries/timeseries.py

```diff
@@ -1476,36 +1489,43 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
         scipy.signal
 
         Notes
-        -----"""
-        # build whitener
+        -----
+        The `window` argument is used in both ASD estimation and FIR filter
+        design.
+
+        Due to filter corruption at the boundaries, a segment of length
+        `fduration` will be automatically cropped from the beginning and
+        end of the output.
+
+        For more on inverse spectrum truncation, see arXiv:gr-qc/0509116
+        """
+        filter_length = int((fduration * self.sample_rate).decompose().value)
         if asd is None:
             asd = self.asd(fftlength, overlap=overlap,
                            method=method, window=window, **kwargs)
-        if isinstance(asd, units.Quantity):
-            asd = asd.value
-        invasd = 1. / asd
-        # build window
-        nfft = int((fftlength * self.sample_rate).decompose().value)
-        noverlap = int((overlap * self.sample_rate).decompose().value)
-        # format window
-        if type(window).__module__ == 'lal.lal':
-            window = window.data.data
-        elif not isinstance(window, numpy.ndarray):
-            window = signal.get_window(window, nfft)
-        # create output series
-        nstride = nfft - noverlap
-        nsteps = 1 + int((self.size - nfft) / nstride)
-        out = numpy.zeros(nsteps * nstride + noverlap).view(type(self))
-        out.__metadata_finalize__(self)
-        out._unit = self.unit
-        del out.times
-        # loop over ffts and whiten each one
-        for i in range(nsteps):
-            x = i * nstride
-            y = x + nfft
-            in_ = self[x:y].detrend(detrend) * window
-            out.value[x:y] += npfft.irfft(in_.fft().value * invasd)
-        return out
+        asd = asd.interpolate(1./self.duration.decompose().value)
+        invasd = 1. / asd.value
+        # zero the edges and apply a hard highpass
+        invasd[0] = 0
+        invasd[asd.size - 1] = 0
+        if highpass:
+            ncorner = int(highpass / asd.df.decompose().value)
+            invasd[0:ncorner] = 0
+        # truncate and window in the time domain
+        tdw = npfft.irfft(invasd)
+        trunc_start = int(filter_length / 2)
+        trunc_stop = self.size - trunc_start
+        window = signal.get_window(window, filter_length)
+        tdw[0:trunc_start] *= window[trunc_start:filter_length]
+        tdw[trunc_stop:self.size] *= window[0:trunc_start]
+        tdw[trunc_start:trunc_stop] = 0
+        # apply truncated zero-phase filter in the frequency domain
+        invasd = (self.duration.value/2) * numpy.abs(npfft.rfft(tdw))
+        in_ = self.detrend(detrend)
+        out = type(self)(npfft.irfft(in_.fft().value * invasd))
+        out.__array_finalize__(self)
+        return out.crop(out.t0.decompose().value + fduration,
+                        (out.t0 + out.duration).decompose().value - fduration)
```

Comment:
it would be cleaner here to use `out.xspan`, which does all decompositions internally, something like

```python
return out.crop(*out.xspan.contract(fduration))
```

Reply:
I like that a lot better as well! This is fixed in commit 66e9982.